src/tool/hpcrun/utilities/linuxtimer.c

```diff
@@ -0,0 +1,177 @@
+//
+// * Neither the name of Rice University (RICE) nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// This software is provided by RICE and contributors "as is" and any
+// express or implied warranties, including, but not limited to, the
+// implied warranties of merchantability and fitness for a particular
+// purpose are disclaimed. In no event shall RICE or contributors be
+// liable for any direct, indirect, incidental, special, exemplary, or
+// consequential damages (including, but not limited to, procurement of
+// substitute goods or services; loss of use, data, or profits; or
+// business interruption) however caused and on any theory of liability,
+// whether in contract, strict liability, or tort (including negligence
+// or otherwise) arising in any way out of the use of this software, even
+// if advised of the possibility of such damage.
+//
+// ******************************************************* EndRiceCopyright *
+
+//*****************************************************************************
+// system includes
+//*****************************************************************************
+
+#include <signal.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+
+
+
+//*****************************************************************************
+// local includes
+//*****************************************************************************
+
+#include "linuxtimer.h"
+
+
+
+//*****************************************************************************
+// local includes
+//*****************************************************************************
+
+// the man pages cite sigev_notify_thread_id in struct sigevent,
+// but often the only name is a hidden union name.
+#ifndef sigev_notify_thread_id
+#define sigev_notify_thread_id  _sigev_un._tid
+#endif
+
+
+
+//*****************************************************************************
+// private operations
+//*****************************************************************************
+
+static void
+timespec_set
+(
+ struct timespec *time,
+ time_t sec,
+ long nsec
+)
+{
+  time->tv_sec = sec;
+  time->tv_nsec = nsec;
+}
+
+
+
+//*****************************************************************************
+// interface operations
+//*****************************************************************************
+
+int
+linuxtimer_create
+(
+ linuxtimer_t *td,
+ clockid_t clock,
+ int signal
+)
+{
+  memset(&td->sigev, 0, sizeof(td->sigev));
+
+  td->sigev.sigev_notify = SIGEV_THREAD_ID;
+  td->sigev.sigev_signo = signal;
+  td->sigev.sigev_value.sival_ptr = &td->timerid;
+  td->sigev.sigev_notify_thread_id = syscall(SYS_gettid);
+
+  int ret = timer_create(clock, &td->sigev, &td->timerid);
+
+  if (ret) td->timerid = NULL;
+
+  return ret;
+}
+
+
+int
+linuxtimer_newsignal
```

Comment:
@jmellorcrummey Do we want to query whether the next_signal has been used by the application and skip the ones that have been registered?